tracing-effectful/src/Effectful/Tracing.hs

```diff
@@ -18,62 +18,65 @@ module Effectful.Tracing
     ) where
 
 import Effectful
-import Effectful.Dispatch.Static
+import Effectful.Dispatch.Dynamic
+import Effectful.Reader.Static
 import Control.Monad.Trace
 import Control.Monad.Trace.Class
 import Control.Monad.Catch (finally)
 import Monitor.Tracing
 import Monitor.Tracing.Zipkin (Zipkin(zipkinTracer), Settings, new, publish)
 
--- | Provides the ability to send traces to a backend
-data Trace :: Effect
+-- | Provides the ability to send traces to a backend.
+data Trace :: Effect where
+  Trace :: Builder -> m a -> Trace m a
+  ActiveSpan :: Trace m (Maybe Span)
+  AddSpanEntry :: Key -> Value -> Trace m ()
 
-type instance DispatchOf Trace = Static WithSideEffects
-newtype instance StaticRep Trace = Trace (Maybe Scope)
+type instance DispatchOf Trace = Dynamic
 
--- | Run the 'Tracing' effect.
+-- | Run the 'Trace' effect.
 --
--- /Note:/ this is the @effectful@ version of 'runTracingT'.
-runTracing
+-- /Note:/ this is the @effectful@ version of 'runTraceT'.
+runTrace
   :: IOE :> es
-  => Eff (Trace : es) a
-  -> Tracer
+  => Tracer
+  -> Eff (Trace : es) a
   -> Eff es a
-runTracing actn = runTracingMaybe  actn . pure
+runTrace tracer = reinterpret (runReader initialScope) $ \env -> \case
+  Trace bldr action -> localSeqUnlift env $ \unlift -> do
+    scope <- ask
+    withSeqEffToIO $ \runInIO -> do
+      traceWith bldr scope $ \childScope -> do
+        runInIO . local (const childScope) $ unlift action
+  ActiveSpan -> asks scopeSpan
+  AddSpanEntry key value -> do
+    scope <- ask
+    liftIO $ addSpanEntryWith (Just scope) key value
+  where
+    initialScope = Scope tracer Nothing Nothing Nothing
 
-runTracingMaybe
-  :: IOE :> es
-  => Eff (Trace : es) a
-  -> Maybe Tracer
-  -> Eff es a
-runTracingMaybe actn mbTracer = 
-  let scope = fmap (\tracer -> Scope tracer Nothing Nothing Nothing) mbTracer
-  in evalStaticRep (Trace scope) actn
+-- | Run the 'Trace' effect with a dummy handler that does no tracing.
+runNoTrace :: IOE :> es => Eff (Trace : es) a -> Eff es a
+runNoTrace = interpret $ \env -> \case
+  Trace _ action -> localSeqUnlift env $ \unlift -> unlift action
+  ActiveSpan -> pure Nothing
+  AddSpanEntry _ _ -> pure ()
 
--- | Convenience method to start a 'Zipkin', run an action, and publish all spans before returning.
+-- | Convenience method to start a 'Zipkin', run an action, and publish all
+-- spans before returning.
 withZipkin :: (IOE :> es) => Settings -> (Zipkin -> Eff es a) -> Eff es a
 withZipkin settings f = do
-  zipkin <- unsafeEff_ $ new settings
+  zipkin <- liftIO $ new settings
```

Comment:
For my general understanding of effectful: are `liftIO` and `unsafeEff_` exchangeable in general (or at least when `es` is mostly `IOE` ?) or is this a specific case ?

Reply:
Not quite - `liftIO` requires `IOE` effect to be in scope and `unsafeEff_` doesn't - that's why it's called unsafe ;)

It's an escape hatch mostly for statically dispatched effect operations which shouldn't require `IOE` (only the runner should), but I was unable to find a more satisfying solution to this problem so far.